widget/entry-with-completion.go

```diff
@@ -0,0 +1,127 @@
+package widget
+
+import (
+	"fyne.io/fyne/v2"
+	"fyne.io/fyne/v2/widget"
+)
+
+// CompletionEntry is an Entry with options displayed in a PopUpMenu.
+type CompletionEntry struct {
+	widget.Entry
+	menu      *fyne.Menu
+	popupMenu *widget.PopUpMenu
+	items     []string
+	pause     bool
+}
+
+// NewCompletionEntry creates a new CompletionEntry which creates a popup menu that responds to keystrokes to navigate through the items without losing the editing ability of the text input..
+func NewCompletionEntry(options []string) *CompletionEntry {
+	c := &CompletionEntry{items: options}
+	c.ExtendBaseWidget(c)
+	// c.Entry = widget.NewEntry()
+	return c
+}
+
+// ShowCompletion displays the completion menu
+func (c *CompletionEntry) ShowCompletion() {
+	if c.pause {
+		return
+	}
+	items := make([]*fyne.MenuItem, len(c.items))
+	for i, item := range c.items {
+		s := item // dereferencing
+		items[i] = fyne.NewMenuItem(s, func() {
+			c.setTextFromMenu(s)
+			c.popupMenu.Hide()
+		})
+	}
+	if c.popupMenu != nil {
+		c.popupMenu.Hide()
+	}
+	// New menu
+	c.menu = fyne.NewMenu("", items...)
+	holder := fyne.CurrentApp().Driver().CanvasForObject(c)
+	c.popupMenu = widget.NewPopUpMenu(c.menu, holder)
+	c.bindKey(holder)
+
+	pos := fyne.CurrentApp().Driver().AbsolutePositionForObject(c)
+	c.popupMenu.ShowAtPosition(fyne.Position{X: pos.X, Y: pos.Y + c.Size().Height})
+}
+
+// HideCompletion hides the completion menu.
+func (c *CompletionEntry) HideCompletion() {
+	if c.popupMenu != nil {
+		c.popupMenu.Hide()
+	}
+}
+
+// SetOptions set the completion list with itemList.
+func (c *CompletionEntry) SetOptions(itemList []string) {
+	c.items = itemList
+}
+
+// Prevent the menu to open when the user validate value from the menu.
+func (c *CompletionEntry) setTextFromMenu(s string) {
+	c.pause = true
+	c.Entry.SetText(s)
+	c.Entry.Refresh()
+	c.pause = false
+}
+
+// FocusGained make the Entry to be focused.
+//
+// Implements: fyne.Focusable
+func (c *CompletionEntry) FocusGained() {
```

Comment:
When you embed a type, you don't need to redefine its functions, unless you need to change its behavior.